plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+	}
+	GList *temp = sdp->m_lines;
+	while(temp) {
+		janus_sdp_mline *m = (janus_sdp_mline *)temp->data;
+		session->media.require_srtp = session->media.require_srtp || (m->proto && !strcasecmp(m->proto, "RTP/SAVP"));
+		if(m->type == JANUS_SDP_AUDIO) {
+			if(m->port) {
+				if(m->port != session->media.remote_audio_rtp_port) {
+					/* This is an update and an address changed */
+					if(changed)
+						*changed = TRUE;
+				}
+				session->media.has_audio = 1;
+				session->media.remote_audio_rtp_port = m->port;
+				session->media.remote_audio_rtcp_port = m->port+1;	/* FIXME We're assuming RTCP is on the next port */
+				if(m->direction == JANUS_SDP_SENDONLY || m->direction == JANUS_SDP_INACTIVE)
+					session->media.audio_send = FALSE;
+				else
+					session->media.audio_send = TRUE;
+			} else {
+				session->media.audio_send = FALSE;
+			}
+		} else if(m->type == JANUS_SDP_VIDEO) {
+			if(m->port) {
+				if(m->port != session->media.remote_video_rtp_port) {
+					/* This is an update and an address changed */
+					if(changed)
+						*changed = TRUE;
+				}
+				session->media.has_video = 1;
+				session->media.remote_video_rtp_port = m->port;
+				session->media.remote_video_rtcp_port = m->port+1;	/* FIXME We're assuming RTCP is on the next port */
+				if(m->direction == JANUS_SDP_SENDONLY || m->direction == JANUS_SDP_INACTIVE)
+					session->media.video_send = FALSE;
+				else
+					session->media.video_send = TRUE;
+			} else {
+				session->media.video_send = FALSE;
+			}
+		} else {
+			JANUS_LOG(LOG_WARN, "Unsupported media line (not audio/video)\n");
+			temp = temp->next;
+			continue;
+		}
+		if(m->c_addr) {
+			if(update && strcmp(m->c_addr, session->media.remote_ip)) {
+				/* This is an update and an address changed */
+				if(changed)
+					*changed = TRUE;
+			}
+			g_free(session->media.remote_ip);
+			session->media.remote_ip = g_strdup(m->c_addr);
+		}
+		if(update) {
+			/* FIXME This is a session update, we only accept changes in IP/ports */
+			temp = temp->next;
+			continue;
+		}
+		GList *tempA = m->attributes;
+		while(tempA) {
+			janus_sdp_attribute *a = (janus_sdp_attribute *)tempA->data;
+			if(a->name) {
+				if(!strcasecmp(a->name, "crypto")) {
+					if(m->type == JANUS_SDP_AUDIO || m->type == JANUS_SDP_VIDEO) {
+						gint32 tag = 0;
+						int suite;
+						char crypto[81];
+						/* FIXME inline can be more complex than that, and we're currently only offering SHA1_80 */
+						int res = sscanf(a->value, "%"SCNi32" AES_CM_128_HMAC_SHA1_%2d inline:%80s",
+							&tag, &suite, crypto);
+						if(res != 3) {
+							JANUS_LOG(LOG_WARN, "Failed to parse crypto line, ignoring... %s\n", a->value);
+						} else {
+							gboolean video = (m->type == JANUS_SDP_VIDEO);
+							int current_suite = video ? session->media.video_srtp_suite_in : session->media.audio_srtp_suite_in;
+							if(current_suite == 0) {
+								if(video)
+									session->media.video_srtp_suite_in = suite;
+								else
+									sessi
```

Comment:
That's how the SIP plugin already worked, and I just migrated (as in copied/pasted) it in here as well. I guess we might indeed have a single thread that polls on all the file descriptors for all sessions and handles all incoming traffic accordingly, but that would be a non trivial refactoring as it is. Besides, there's code (still unused) to handle session updates that could keep the thread busy, and I'm afraid this would impact other sessions.

Anyway, your point on threads in general is valid, and applies to the Janus core as well.